### PR TITLE
Require Rust 2021 edition

### PR DESCRIPTION
**Pull request by mgeisler**
This allows us to clean up and simplify the code.

The changes were originally written by @Dylan-DPC in #1831. This PR fixes a few build errors in that PR, thanks to the helpful comments made by @ISSOtm. I've also rebased the changes onto the latest `master`.

### CODE REVIEW

**Review by ISSOtm — APPROVED** [review]

**Comment by Dylan-DPC** [comment]
@ehuss do we want to do this in a new major version?

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -31,8 +31,7 @@ jobs:
             rust: stable
           - build: msrv
             os: ubuntu-latest
-            # sync MSRV with docs: guide/src/guide/installation.md
-            rust: 1.54.0
+            rust: 1.56.0
```
Comment (ehuss) [inline .github/workflows/main.yml:34]: It looks like there was a comment removed from here. Can we make sure it is retained?

**Comment by ehuss** [comment]
> @ehuss do we want to do this in a new major version?

No, I think a normal release is fine.

File: .github/workflows/main.yml (hunk 1 above)
Comment (mgeisler) [reply .github/workflows/main.yml:35]: Definitely, it should be fixed now.

**Review by ehuss — APPROVED** [review]
Thanks!

**Pull request merged**